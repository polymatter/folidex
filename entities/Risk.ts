/**
 * This file represents a Business Entity in Bob Martins CLEAN architecture {@link https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html}
 * 
 * Responsible for defining the business rules relating to what a risk is.
 * - Types relating to risks are defined here and exported in the top-level types.tsx for other files to pickup
 * - Anything that affects the Risk should call a function here.
 */


/** 
 * The public interface for risks.
 * 
*/
export default interface Risk {
  getId: () => string;
  getLevel: () => RiskLevel;
  getLabel: () => string;
  getMitigation: () => string;
  getContingency: () => string;
  getImpact: () => string;
  getLikelihood: () => string;
  toJson: () => Readonly<RiskProps>;
}

export type RiskLevel = "Low" | "Medium" | "High";

export interface RiskProps {
  id: string;
  level: RiskLevel;
  label: string;
  mitigation: string;
  contingency: string;
  impact: string;
  likelihood: string;
}

export type RiskSummaryProps = Pick<RiskProps, 'level' | 'label'>

export type RiskSummary = Pick<Risk, 'getLevel' | 'getLabel'> & { toJson: () => Readonly<RiskSummaryProps> }

export interface BuildMakeRiskProps {
}

export function buildMakeRisk({ }: BuildMakeRiskProps) {
  /** Responsibility for creating the only function that can create risks. arguments are dependencies to be injected in */
  return function makeRisk({ id, level, label, mitigation, contingency, impact, likelihood }: RiskProps): Readonly<Risk> {
    // validation of properties throwing exceptions if invalid
    return Object.freeze({
      getId: () => id,
      getLevel: () => level,
      getLabel: () => label,
      getMitigation: () => mitigation,
      getContingency: () => contingency,
      getImpact: () => impact,
      getLikelihood: () => likelihood,
      toJson: () => Object.freeze({ id, level, label, mitigation, contingency, impact, likelihood }),
    })
  }
}

export interface BuildMakeRiskSummaryProps {
}

export function buildMakeRiskSummary({ }: BuildMakeRiskSummaryProps) {
  return function makeRiskSummary({ level, label }: RiskSummaryProps): Readonly<RiskSummary> {
    return Object.freeze({
      getLevel: () => level,
      getLabel: () => label,
      toJson: () => Object.freeze({ level, label })
    })
  }
}
export interface BuildMakeNewRiskProps extends BuildMakeRiskProps {
  Id: { generateId: () => string }
}

export function buildMakeNewRisk({ Id, ...others }: BuildMakeNewRiskProps) {
  return function makeNewRisk(riskProps: Omit<RiskProps, 'id'>): Readonly<Risk> {
    const makeRisk = buildMakeRisk(others);
    const id = Id.generateId();
    return makeRisk({ ...riskProps, id });
  }
}