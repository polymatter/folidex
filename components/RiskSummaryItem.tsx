import React from 'react'
import { StyleSheet } from 'react-native';
import { RiskSummary } from '../entities/Risk';
import { Text, View } from './Themed';
import RiskLevelBadge from './RiskLevelBadge';

const RiskSummaryItem = ({ risk }: { risk: RiskSummary }) => {

  return (
    <View style={styles.container}>
      <Text style={styles.label}>
        <RiskLevelBadge level={risk.level} />
        {risk.label}
      </Text>
      <View style={styles.separator} lightColor="#aaa" darkColor="rgba(255,255,255,0.1)" />
    </View>
  )
}

const styles = StyleSheet.create({
  label: {
    display: 'flex',
    flexGrow: 10
  },
  container: {
    flex: 1,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    marginVertical: 10,
    height: 1,
    width: '90%',
  },
});

export default RiskSummaryItem
