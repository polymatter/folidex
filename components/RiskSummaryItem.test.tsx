import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import RiskSummaryItem from './RiskSummaryItem';
import RiskStoreContext from '../store/RiskStore';
import mockdata from '../mocks/RiskStore.json';
import Risk from '../entities/Risk';

describe('<RiskSummaryItem>', () => {

  test('can render a risk item', () => {

    const tree = render(
      <RiskStoreContext.Provider value={mockdata as Risk[]}>
        <RiskSummaryItem riskId={"Id_1"} />
      </RiskStoreContext.Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
})