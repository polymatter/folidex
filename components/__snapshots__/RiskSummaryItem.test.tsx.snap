// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RiskSummaryItem> can render a risk item 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {},
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "lineHeight": 10,
            "minHeight": 30,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
        ]
      }
    >
      <Icon
        color="#dc3545"
        name="dangerous"
        size={24}
      />
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            Array [
              Object {
                "borderRadius": 15,
                "display": "flex",
                "flexShrink": 0,
                "fontSize": 12,
                "marginRight": 30,
                "paddingLeft": 8,
                "paddingRight": 8,
              },
              Object {
                "color": "#dc3545",
              },
            ],
          ]
        }
      >
        High
      </Text>
    </View>
    The server capacity initially defined may be inadequate
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#aaa",
        },
        Object {
          "height": 1,
          "marginVertical": 10,
          "width": "90%",
        },
      ]
    }
  />
</View>
`;
